File: sdk/swift/Sources/SeocheonSDK/Infrastructure/Signing/DirectService.swift
import Foundation

/// Signs transactions using a mnemonic directly (test/development only).
public final class DirectSigningService: SigningService, @unchecked Sendable {
    private let privKey: PrivateKey
    private let address: String
    private let pubKeyData: Data

    public init(mnemonic: String) throws {
        guard !mnemonic.isEmpty else {
            throw SDKError.signingFailed("mnemonic is required")
        }
        self.privKey = try Bip44.deriveKeyFromMnemonic(mnemonic)
        self.pubKeyData = privKey.pubKey
        self.address = try Address.fromPubKey(pubKeyData)
    }

    public func sign(_ txBytes: Data) async throws -> Data {
        return try privKey.sign(txBytes)
    }

    public func getAddress() -> String {
        return address
    }

    public func getPubKey() -> Data {
        return pubKeyData
    }
}

File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/infrastructure/signing/DirectService.kt
package com.seocheon.sdk.infrastructure.signing

import com.seocheon.sdk.internal.crypto.Address
import com.seocheon.sdk.internal.crypto.Bip44
import com.seocheon.sdk.internal.crypto.PrivateKey

/**
 * Signs transactions using a mnemonic directly (test/development only).
 */
class DirectService(mnemonic: String) : SigningService {

    private val privKey: PrivateKey
    private val address: String
    private val pubKey: ByteArray

    init {
        require(mnemonic.isNotBlank()) { "mnemonic is required" }
        privKey = Bip44.deriveKeyFromMnemonic(mnemonic)
        pubKey = privKey.pubKey()
        address = Address.fromPubKey(pubKey)
    }

    override suspend fun sign(txBytes: ByteArray): ByteArray = privKey.sign(txBytes)

    override fun getAddress(): String = address

    override fun getPubKey(): ByteArray = pubKey
}
